crates/ol/stf/src/stf.rs

```diff
@@ -0,0 +1,600 @@
+};
+use strata_snark_acct_sys as snark_sys;
+use strata_snark_acct_types::{SnarkAccountUpdateContainer, UpdateOperationData};
+
+use crate::{
+    ExecOutput,
+    asm::process_asm_log,
+    context::BlockExecContext,
+    error::{StfError, StfResult},
+    ledger::LedgerInterfaceImpl,
+    post_exec_block_validate, pre_exec_block_validate,
+};
+
+/// Executes an OL block with full validation.
+///
+/// Performs pre-execution validation (header checks), executes transactions,
+/// handles epoch sealing for terminal blocks, and validates post-state root.
+///
+/// Returns execution output containing the new state root and accumulated logs.
+pub fn execute_block<S: StateAccessor>(
+    ctx: BlockExecContext,
+    state_accessor: &mut S,
+    block: OLBlock,
+) -> StfResult<ExecOutput> {
+    // Do some pre execution validation
+    pre_exec_block_validate(&block, ctx.prev_header(), ctx.params())
+        .map_err(StfError::BlockValidation)?;
+
+    let exec_output = execute_block_body(ctx, state_accessor, block.body())?;
+
+    // Post execution block validation. Checks state root and logs root.
+    post_exec_block_validate::<S>(&block, exec_output.state_root(), exec_output.logs())
+        .map_err(StfError::BlockValidation)?;
+
+    Ok(exec_output)
+}
+
+/// Executes block body without any block level validation.
+///
+/// Used by block assembly where validation isn't needed. Executes all transactions
+/// and handles epoch sealing but skips header and state root validation.
+pub fn execute_block_body<S: StateAccessor>(
+    ctx: BlockExecContext,
+    state_accessor: &mut S,
+    block_body: &OLBlockBody,
+) -> StfResult<ExecOutput> {
+    // Execute transactions.
+    execute_transactions(&ctx, state_accessor, block_body.txs())?;
+
+    // Check if needs to seal epoch, i.e is a terminal block.
+    if let Some(l1update) = block_body.l1_update() {
+        let preseal_root = state_accessor.compute_state_root();
+
+        // Check pre_seal_root matches with l1update preseal_root.
+        if l1update.preseal_state_root() != preseal_root {
+            return Err(StfError::PresealRootMismatch {
+                expected: l1update.preseal_state_root(),
+                got: preseal_root,
+            });
+        }
+        seal_epoch(&ctx, state_accessor, l1update)?;
+    }
+
+    let new_root = state_accessor.compute_state_root();
+    let out = ExecOutput::new(new_root, ctx.into_logs());
+    Ok(out)
+}
+
+/// Executes the OL transactions and updates the state accordingly.
+pub fn execute_transactions(
+    ctx: &BlockExecContext,
+    state_accessor: &mut impl StateAccessor,
+    txs: &[OLTransaction],
+) -> StfResult<()> {
+    for tx in txs {
+        // validate tx extra
+        validate_tx_extra(state_accessor, tx.extra())?;
+        execute_transaction(ctx, state_accessor, tx)?;
+    }
+    Ok(())
+}
+
+pub fn validate_tx_extra(
+    state_accessor: &impl StateAccessor,
+    extra: &TransactionExtra,
```

Comment:
Can you clarify more about the extra data? I see that it was already introduced [here](https://github.com/alpenlabs/alpen/pull/1085/commits/8d9aa6e0742f6621ada5b065c22b161e36b8e843), but it's not clear to me what exactly this is needed.